firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/messages.txt

```diff
@@ -0,0 +1,181 @@
+    WHEN("The internal type member is set to an arbitrary value and then write method is called")
+    // setup:
+    // type is set to 4
+    // A sensor measurements payload message is set, which has value of type as 2
+      THEN("The write method reports ok status")
+      THEN("After the write method is called, type member remains unchanged)
+
+
+  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 126 bytes")
+    WHEN("The payload is a sensor measurements message and write method is called")
+      THEN("The write method reports ok status")
+      THEN("The body's header correctly stores the value of message type obtained from the message payload tag")
+      THEN("The output buffer is as expected")
+
+  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 508 bytes")
+    WHEN("The sensor measurments message data is written")
+      THEN("The write method reports ok status")
+      THEN("The body's header correctly stores the value of message type obtained from the message payload tag")
+      THEN("The output buffer is as expected")
+
+/* Write function for each of the message types (example: sensor_measurements)
+    WHEN("The payload is a sensor measurements message and write method is called")
+    // setup: sensor_measurements.flow = 50
+      THEN("The write method reports ok status")
+      THEN("The body's header correctly stores the value of message type obtained from the message payload tag")
+      THEN("The output buffer is as expected")
+/*
+
+/* Parse function
+
+Scenario: Message: The parse method correctly updates internal message type and sets the message payload fields after decoding from the input buffer
+
+  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes")
+
+    WHEN("A body with an encoded payload and 1 byte header of value equal to 0 is parsed")
+      THEN("The parse method reports invalid type status")
+      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
+      THEN("The payload Taggedunion tag is set to Unknown")
+      THEN("The message payload fields are not set")
+
+    WHEN("An empty input buffer body is parsed")
+      THEN("The parse method reports invalid length status")
+      THEN("The payload Taggedunion tag is set to Unknown")
+      THEN("The message payload fields are not set")
+
+    WHEN("A body with an empty payload and 1 byte header of value equal to the message descriptor array size is parsed")
+      THEN("The parse method reports invalid type status")
+      THEN("After the parse method is called, The value assigned to the type member is equal to the input_buffer body's header")
+      THEN("The payload Taggedunion tag is set to Unknown")
```

Comment:
There are three ways `parse` can return `invalid_type`; in one of them, it does not change `payload.tag`, and in the other two, it depends on the GIVEN/WHEN (this will vary depending on the `MessageTypes` template parameter and the `pb_protobuf_descriptors` argument). We should cover each of the three ways `invalid_type` can be returned. It could make sense for this to be a separate scenario, because it may involve multiple GIVENs.

Reply:
Where is the test of `invalid_type` resulting from `!MessageTypes::includes(type)`?

Reply:
So the tests where the body with `header` value greater than and equal to descriptor array size (8), both actually failed this test instead, as `MessageTypes` enum has enums till 7. Therefore I have re-structured these tests now